packages/permission-controller/src/PermissionController.ts

```diff
@@ -1991,6 +2110,214 @@ export class PermissionController<
     }
   }
 
+  /**
+   * Merges a set of incrementally requested permissions into the existing permissions of
+   * the requesting subject. The merge is a right-biased union, where the existing
+   * permissions are the left-hand side, and the incrementally requested permissions are
+   * the right-hand side.
+   *
+   * @param existingPermissions - The subject's existing permissions.
+   * @param incrementalRequestedPermissions - The requested permissions to merge.
+   * @returns The merged permissions and the resulting diff.
+   */
+  #mergeIncrementalPermissions(
+    existingPermissions: Exclude<
+      ReturnType<typeof this.getPermissions>,
+      undefined
+    >,
+    incrementalRequestedPermissions: RequestedPermissions,
+  ):
+    | [
+        SubjectPermissions<
+          ValidPermission<string, ExtractCaveats<ControllerCaveatSpecification>>
+        >,
+        PermissionDiffMap<string, CaveatConstraint>,
+      ]
+    | [] {
+    const permissionDiffMap: PermissionDiffMap<string, CaveatConstraint> = {};
+
+    // Use immer's produce as a convenience for calculating the new permissions
+    // without mutating the existing permissions or committing the results to state.
+    const newPermissions = immerProduce(
+      existingPermissions,
+      (draftExistingPermissions) => {
+        const leftPermissions =
+          draftExistingPermissions as RequestedPermissions;
+
+        Object.entries(incrementalRequestedPermissions).forEach(
+          ([targetName, rightPermission]) => {
+            if (
```

Comment:
Wondering if we could simplify the branching here.

Would you hate removing the extra branch in favor of
```
const [newPermission, caveatsDiff] = this.#mergePermission(
  leftPermissions.[targetName] ?? {},
  rightPermission,
);

if (Object.keys(caveatsDiff).length > 0) {
  leftPermissions[targetName] = newPermission;
  permissionDiffMap[targetName] = caveatsDiff;
}
```

Reply:
This sparks joy (with minor modifications): a7d6cd5a